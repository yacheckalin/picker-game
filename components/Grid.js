import React, { useEffect } from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

const StyledCell = styled.div`
  margin: 0px;
  padding: 0px;
  width: ${(prop) => prop.width}px;
  height: ${(prop) => prop.height}px;
  margin: 0px;
  padding: 0px;
  box-sizing: border-box;
  background-color: ${(prop) => prop.color};
  display: inline-flexbox;
  position: relative;
`;

const StyledRow = styled.div`
  margin: 0px;
  padding: 0px;
`;

const Grid = ({ data, cellSize, recalculate }) => {
  useEffect(() => {
    // calculate the pointer height, width size from the first element on a grid
    const el = document.getElementById("#grid-cell-0-0");
    let offsetHeight, offsetWidth;
    if (el) {
      offsetHeight = el.offsetHeight;
      offsetWidth = el.offsetWidth;
    } else {
      offsetHeight = 0;
      offsetWidth = 0;
    }
    recalculate({ pointerWidth: offsetWidth, pointerHeight: offsetHeight });
  });
  return (
    <div>
      {data.map((row, y) => (
        <StyledRow key={y}>
          {row.map((col, x) => (
            <StyledCell
              className={`grid-cell ${
                col == 8 ? "light-green lighten-4" : "blue"
              }`}
              key={x}
              width={cellSize}
              height={cellSize}
              id={`#grid-cell-${y}-${x}`}
            >
              &nbsp;
            </StyledCell>
          ))}
        </StyledRow>
      ))}
    </div>
  );
};

Grid.propTypes = {
  data: PropTypes.array.isRequired,
  cellSize: PropTypes.number.isRequired,
  recalculate: PropTypes.func,
};

export default Grid;
