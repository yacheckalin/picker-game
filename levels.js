import { KEYS, DOORS, WALL, VISITED } from "./constants";

export const LEVELS = [
  [
    "level_1",
    [
      [VISITED, 0, WALL, WALL, 1, 1, 1, 1],
      [1, 0, 1, 0, 0, 0, KEYS.RED_KEY, 1],
      [1, 0, 1, 0, 1, 1, 1, 1],
      [1, 0, 1, 0, 1, 0, KEYS.BLUE_KEY, 1],
      [1, 0, 0, 0, 1, 0, 0, 1],
      [1, 0, 1, 0, 1, 0, 0, 1],
      [1, KEYS.GREEN_KEY, 1, 0, 0, 0, 0, DOORS.BLUE_DOOR],
      [1, 1, 1, 1, 1, 1, 1, 1],
    ],
  ],
  [
    "level_2",
    [
      [VISITED, 0, WALL, WALL, 1, 1, 1, 1],
      [1, 0, 1, 0, 0, 0, 0, 1],
      [1, 0, 1, 0, 1, 1, 1, 1],
      [1, 0, 1, 0, 1, 0, KEYS.BLUE_KEY, 1],
      [1, 0, 0, 0, 1, 0, 0, 1],
      [1, 0, 1, 0, 1, 0, 0, 1],
      [1, 0, 1, 0, 0, 0, 0, DOORS.BLUE_DOOR],
      [1, 1, 1, 1, 1, 1, 1, 1],
    ],
  ],
];
