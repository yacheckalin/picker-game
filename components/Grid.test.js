import React from "react";
import Grid from "./Grid";
import { mount } from "enzyme";
import toJSON from "enzyme-to-json";

import { LEVELS } from "../levels";

describe("<Grid />", () => {
  const props = {
    data: LEVELS[0][1],
    cellSize: 40,
    recalculate: jest.fn(),
  };
  test("renders and displays properly", () => {
    const wrapper = mount(<Grid {...props} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test("all grid cell drawed properly", () => {
    const wrapper = mount(<Grid {...props} />);
    expect(wrapper.find("div.grid-cell")).toHaveLength(16);
  });
});
