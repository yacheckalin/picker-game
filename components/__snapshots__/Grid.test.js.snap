// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> renders and displays properly 1`] = `
<Grid
  cellSize={40}
  data={
    Array [
      Array [
        8,
        0,
        1,
        0,
      ],
      Array [
        0,
        1,
        0,
        Array [
          90,
          "BLUE",
        ],
      ],
      Array [
        Array [
          20,
          "BLUE",
        ],
        1,
        0,
        0,
      ],
      Array [
        0,
        0,
        0,
        Array [
          20,
          "BLUE",
        ],
      ],
    ]
  }
  recalculate={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "pointerHeight": 0,
            "pointerWidth": 0,
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div>
    <styled.div
      key="0"
    >
      <div
        className="sc-AxirZ geALnH"
      >
        <styled.div
          className="grid-cell light-green lighten-4"
          height={40}
          id="#grid-cell-0-0"
          key="0"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell light-green lighten-4"
            height={40}
            id="#grid-cell-0-0"
            width={40}
          >
             
          </div>
        </styled.div>
        <styled.div
          className="grid-cell blue"
          height={40}
          id="#grid-cell-0-1"
          key="1"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell blue"
            height={40}
            id="#grid-cell-0-1"
            width={40}
          >
             
          </div>
        </styled.div>
        <styled.div
          className="grid-cell blue"
          height={40}
          id="#grid-cell-0-2"
          key="2"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell blue"
            height={40}
            id="#grid-cell-0-2"
            width={40}
          >
             
          </div>
        </styled.div>
        <styled.div
          className="grid-cell blue"
          height={40}
          id="#grid-cell-0-3"
          key="3"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell blue"
            height={40}
            id="#grid-cell-0-3"
            width={40}
          >
             
          </div>
        </styled.div>
      </div>
    </styled.div>
    <styled.div
      key="1"
    >
      <div
        className="sc-AxirZ geALnH"
      >
        <styled.div
          className="grid-cell blue"
          height={40}
          id="#grid-cell-1-0"
          key="0"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell blue"
            height={40}
            id="#grid-cell-1-0"
            width={40}
          >
             
          </div>
        </styled.div>
        <styled.div
          className="grid-cell blue"
          height={40}
          id="#grid-cell-1-1"
          key="1"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell blue"
            height={40}
            id="#grid-cell-1-1"
            width={40}
          >
             
          </div>
        </styled.div>
        <styled.div
          className="grid-cell blue"
          height={40}
          id="#grid-cell-1-2"
          key="2"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell blue"
            height={40}
            id="#grid-cell-1-2"
            width={40}
          >
             
          </div>
        </styled.div>
        <styled.div
          className="grid-cell blue"
          height={40}
          id="#grid-cell-1-3"
          key="3"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell blue"
            height={40}
            id="#grid-cell-1-3"
            width={40}
          >
             
          </div>
        </styled.div>
      </div>
    </styled.div>
    <styled.div
      key="2"
    >
      <div
        className="sc-AxirZ geALnH"
      >
        <styled.div
          className="grid-cell blue"
          height={40}
          id="#grid-cell-2-0"
          key="0"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell blue"
            height={40}
            id="#grid-cell-2-0"
            width={40}
          >
             
          </div>
        </styled.div>
        <styled.div
          className="grid-cell blue"
          height={40}
          id="#grid-cell-2-1"
          key="1"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell blue"
            height={40}
            id="#grid-cell-2-1"
            width={40}
          >
             
          </div>
        </styled.div>
        <styled.div
          className="grid-cell blue"
          height={40}
          id="#grid-cell-2-2"
          key="2"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell blue"
            height={40}
            id="#grid-cell-2-2"
            width={40}
          >
             
          </div>
        </styled.div>
        <styled.div
          className="grid-cell blue"
          height={40}
          id="#grid-cell-2-3"
          key="3"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell blue"
            height={40}
            id="#grid-cell-2-3"
            width={40}
          >
             
          </div>
        </styled.div>
      </div>
    </styled.div>
    <styled.div
      key="3"
    >
      <div
        className="sc-AxirZ geALnH"
      >
        <styled.div
          className="grid-cell blue"
          height={40}
          id="#grid-cell-3-0"
          key="0"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell blue"
            height={40}
            id="#grid-cell-3-0"
            width={40}
          >
             
          </div>
        </styled.div>
        <styled.div
          className="grid-cell blue"
          height={40}
          id="#grid-cell-3-1"
          key="1"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell blue"
            height={40}
            id="#grid-cell-3-1"
            width={40}
          >
             
          </div>
        </styled.div>
        <styled.div
          className="grid-cell blue"
          height={40}
          id="#grid-cell-3-2"
          key="2"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell blue"
            height={40}
            id="#grid-cell-3-2"
            width={40}
          >
             
          </div>
        </styled.div>
        <styled.div
          className="grid-cell blue"
          height={40}
          id="#grid-cell-3-3"
          key="3"
          width={40}
        >
          <div
            className="sc-AxjAm iYVETC grid-cell blue"
            height={40}
            id="#grid-cell-3-3"
            width={40}
          >
             
          </div>
        </styled.div>
      </div>
    </styled.div>
  </div>
</Grid>
`;
